### PR TITLE
Extend abilities via gradio tools

### PR DESCRIPTION
**Pull request by freddyaboulton**
### Background
AutoGPT can be more powerful it had more commands at its disposal. Some developers may want to give their agent special abilities in order for it to accomplish its goals.

### Changes
Introduces a `tools` module that stores all the tools the developer wants to give to their agent.

A `tool` is anything that conforms to this interface:

```python
@dataclass
class Tool:
    name: str
    description: str
    run: Callable[[str], str]
    args: Dict[str, str]
```

To add a tool, developers implement this interface and modify the list of `TOOLS` list defined in `tools.py`.

For a given run, the developer can specify which tool to use by specifying `AUTOGPT_TOOLS` variable in the .env file as a comma separated list.

Additionally, I've added a couple of tools from [gradio-tools](https://github.com/freddyaboulton/gradio-tools), a library which implements some tools from the thousands of [Gradio](https://github.com/gradio-app/gradio) spaces on [huggingface](https://huggingface.co/spaces).

As more tools are added to gradio-tools, more AutoGPT users will benefit.

### Documentation
Waiting on feedback on API before writing documentation. 

### Sample Output
autogpt_and_gradio_tools_2

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [ ] I have thoroughly tested my changes with multiple different prompts.
- [ ] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

File: autogpt/commands.py [hunk 1]
```diff
@@ -5,6 +5,18 @@
 from googleapiclient.discovery import build
 from googleapiclient.errors import HttpError
 from autogpt.web import browse_website
+from autogpt import browse
```
Comment (freddyaboulton) [inline autogpt/commands.py:8]: This is from a separate PR. Had to add this to be able to run my changes.

**Comment by dschonholtz** [comment]
I'm trying to understand how this would interact with the existing plugins/commands the agent has. I'm assuming that you get access to all of the gradio tools and hugging face models here but we keep our web browsing super powers and command line powers?

**Comment by freddyaboulton** [comment]
I definitely could have done a better job of explaining this 😂

I think the benefit of this PR is two fold:

1. A framework for extending AutoGPT with new abilities, which I call Tools. A tool can really be anything as long as you implement the minimal interface so the potential is huge. It can even be a transformers model running locally. And developers can add tools without submitting a PR! They just have to edit tools.py and their local AutoGPT will pick it up. 

2. The addition of some tools that leverage [Huggingface Spaces](https://huggingface.co/spaces). HuggingFace spaces hosts thousands of ML apps that are not accessible via API thanks to the [gradio client library](https://gradio.app/getting-started-with-the-python-client/). My [gradio-tools](https://github.com/freddyaboulton/gradio-tools) library is the start of a curated list of ML apps that are accessible via API that can be used by LLM agents. Spaces has state of the art apps for any kind of task, e.g. text to video, audio transcription, controlnet, text to music, so the benefit of including them in AutoGPT is that developers can enhance their agent’s ability with minimal set up.

**Comment by dschonholtz** [comment]
Gotcha. A couple things there, one we have a plugins system coming that will build a system similar to that interface for all of our commands so they can be configured by the user before app start. I think? This is in the pipeline. Therefore we should probably not attempt to build a directly system with that so we have one standardized plugins system. 

Second the hugging face module usage through spaces is very cool! Would this be subject to the same querying for a model that you see when using the spaces web UI?

Generally, I think the spaces usage is great and enabling that is fantastic, but we should probably keep that separate from a generic tool interface as that is a separate issue. 

@BillSchumacher or @Torantulino if you want to weigh in that’d be great.

**Review by nponeccop — APPROVED** [review]

**Comment by nponeccop** [comment]
@freddyaboulton There are conflicts now

**Comment by freddyaboulton** [comment]
@dschonholtz I totally understand what your point about not wanting to add a framework that potentially interferes with the future plugins system. I think part of the value of this PR is that you can add/detract to the set of spaces your agent can call without manually having to add commands to command.py. I wonder if we can merge this in as is and I'd be happy to refactor to conform to the future plugins system when that's ready?

**Comment by dschonholtz** [comment]
@freddyaboulton I actually do not have merge powers here. 
@nponeccop What are your thoughts on this? What would you need to see in order for this to go in assuming merge conflicts to go in? I do think there are a couple valuable ideas here.

**Comment by nponeccop** [comment]
@dschonholtz I guess the best course would be to wait until plugins are ready and release this as a plugin. That would require some reworking.

**Comment by freddyaboulton** [comment]
Solid. @nponeccop when are the plug-ins expected to land?

**Comment by nponeccop** [comment]
Tonight

**Comment by freddyaboulton** [comment]
Thanks @nponeccop! I'll be ready to do the refactor.

**Comment by ntindle** [comment]
Heyo, lets look at making this a plugin. Put it in a folder that has a descriptive name in `https://github.com/Significant-Gravitas/Auto-GPT-Plugins/tree/master/plugins`

**Comment by freddyaboulton** [comment]
Hi @ntindle @dschonholtz @nponeccop , the plugin PR is here: https://github.com/Significant-Gravitas/Auto-GPT-Plugins/pull/14 ! Thank you

**Pull request closed without merging**